store/v2/flat/root_store.go

```diff
@@ -0,0 +1,638 @@
+			return
+		}
+	}
+	// Apply migrations, then clear old schema and write the new one
+	for _, upgrades := range opts.Upgrades {
+		err = pr.migrate(base, upgrades)
+		if err != nil {
+			return
+		}
+	}
+	schemaWriter := prefixdb.NewPrefixWriter(base.stateTxn, schemaPrefix)
+	it, err := schemaView.Iterator(nil, nil)
+	if err != nil {
+		return
+	}
+	for it.Next() {
+		err = schemaWriter.Delete(it.Key())
+		if err != nil {
+			return
+		}
+	}
+	err = it.Close()
+	if err != nil {
+		return
+	}
+	err = schemaView.Discard()
+	if err != nil {
+		return
+	}
+	for skey, typ := range pr.storeSchema {
+		err = schemaWriter.Set([]byte(skey), []byte{byte(typ)})
+		if err != nil {
+			return
+		}
+	}
+	// The migrated contents and schema are not committed until the next store.Commit
+	ret = &rootStore{
+		Store: base,
+		mem:   mem.NewStore(memdb.NewDB()),
+		tran:  transkv.NewStore(memdb.NewDB()),
+		storeMixin: storeMixin{
+			schema:        pr.storeSchema,
+			listenerMixin: opts.listenerMixin,
+			traceMixin:    opts.traceMixin,
+		},
+	}
+	return
+}
+
+// Applies store upgrades to the DB contents.
+func (pr *prefixRegistry) migrate(store *Store, upgrades types.StoreUpgrades) error {
+	// branch state to allow mutation while iterating
+	branch := cachekv.NewStore(store)
+
+	for _, key := range upgrades.Deleted {
+		sst, ix, err := pr.storeInfo(key)
+		if err != nil {
+			return err
+		}
+		if sst != subStorePersistent {
+			return fmt.Errorf("prefix is for non-persistent substore: %v (%v)", key, sst)
+		}
+		pr.reserved = append(pr.reserved[:ix], pr.reserved[ix+1:]...)
+		delete(pr.storeSchema, key)
+
+		sub := prefix.NewStore(store, []byte(key))
+		subbranch := prefix.NewStore(branch, []byte(key))
+		it := sub.Iterator(nil, nil)
+		for ; it.Valid(); it.Next() {
+			subbranch.Delete(it.Key())
+		}
+		it.Close()
+	}
+	for _, rename := range upgrades.Renamed {
+		sst, ix, err := pr.storeInfo(rename.OldKey)
+		if err != nil {
+			return err
+		}
+		if sst != subStorePersistent {
+			return fmt.Errorf("prefix is for non-persistent substore: %v (%v)", rename.OldKey, sst)
+		}
+		pr.reserved = append(pr.reserved[:ix], pr.reserved[ix+1:]...)
+		delete(pr.storeSchema, rename.OldKey)
+		err = pr.ReservePrefix(types.NewKVStoreKey(rename.NewKey), types.StoreTypePersistent)
+		if err != nil {
+			return err
+		}
+
+		sub := prefix.NewStore(store, []byte(rename.OldKey))
+		subbranch := prefix.NewStore(branch, []byte(rename.NewKey))
+		it := sub.Iterator(nil, nil)
+		for ; it.Valid(); it.Next() {
+			subbranch.Set(it.Key(), it.Value())
+		}
+		it.Close()
+	}
+	branch.Write()
+
+	for _, key := range upgrades.Added {
+		err := pr.ReservePrefix(types.NewKVStoreKey(key), types.StoreTypePersistent)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (rs *rootStore) GetKVStore(key types.StoreKey) types.KVStore {
+	return rs.generic().getStore(key.Name())
+}
+
+func (rs *rootStore) Commit() types.CommitID {
+	id := rs.Store.Commit()
+	rs.mem.Commit()
```

Comment:
@roysc , why LGTM says that this doesn't have any effect?

Reply:
it is indeed a no-op if you see the method on `mem.Store`